### ISSUE TITLE
ColorMatrixFilter render bug on dart2js

### ISSUE THREAD

**Issue by mnordine** [bp74/StageXL#235]
I'm getting some odd color banding when applying a ColorMatrixFilter. It works fine in Dartium. The bug occurs when compiling with dart2js. Note this occurs with both Canvas2D and WebGL renderers.

https://gist.github.com/mnordine/2b10124e0bf9448a1150e5afed0f6dcd

Here's output with Dartium:

![good](https://cloud.githubusercontent.com/assets/1924573/14179925/519d1a0e-f737-11e5-9549-2e198aadb24a.png)

Output after dart2js in Chrome:

![bad](https://cloud.githubusercontent.com/assets/1924573/14179934/5a3087c8-f737-11e5-9c1f-bd5cb1edebe9.png)

This isn't specific to GraphicsGradient either. It occurs with bitmaps as well.

It's also interesting that there's a "marching ants" effect in the bad gradient, between the light blue and dark blue.

**Comment by bp74** [comment]
Hi, thanks for the report. I have to take a look at it, but there are two things in your repro sample you should know. In fact your repro uses the worst case scenario of StageXL :)

Gradients: The Graphics class in WebGL currently only supports solid colors and no gradients or patterns. The reason is that drawing vector shapes in Dart/JavaScript and WebGL is hard. We managed to achieve our first milestone, which is drawing vector shapes with solid colors. Gradients and patterns are on the ToDo list and only supported with the Canvas2D renderer right now.

Filters: This is the exact opposite. Filters are only rendered in real time if the renderer is WebGL. It is impossible to render filters in real time with the Canvas2D renderer. You could still use filters with Canvas2D, but you have to call the applyCache(..) method on the DisplayObject. This will render the filter to a temporary and internal BitmapData statically.

**Comment by mnordine** [comment]
Thanks for taking a look.

> Gradients: The Graphics class in WebGL currently only supports solid colors and no gradients or patterns

Yep, I just used a `Shape` for simplest test case for you. Like I said, this banding occurs on bitmaps as well.

I think my test case is ok, because even though it uses WebGL, it still uses `applyCache()`, so the gradient is rendered correctly (only in Dartium). i.e. `ColorMatrixFilter.apply()` is called.

**Comment by bp74** [comment]
I will definitely look at it, maybe not today but probably tomorrow.

**Comment by mnordine** [comment]
Sorry, I should also mention I'm using StageXL 0.13.5+1, Dart SDK 1.15

**Comment by bp74** [comment]
I found the problem. The ColorMatrixFilter in combination with `applyCache` was optimized with integer calculations, unfortunately this does not work for all ColorMatrixFilters. I changed the implementation which is a little bit slower but works correct in all cases.

I also did a second change because of your repro case. The `applyCache` accepted integer parameters, which is not really necessary because any `num` type should be fine. So you don't have to  do the `.toInt()` call on all parameters.
